unittest2.nim

```diff
@@ -526,19 +560,43 @@ proc shouldRun(currentSuiteName, testName: string): bool =
 
   return false
 
-proc ensureInitialized() =
-  withLock formattersLock:
-    {.gcsafe.}:
-      if formatters.len == 0:
-        formatters = @[OutputFormatter(defaultConsoleFormatter())]
+proc cleanupFormatters() {.noconv.} =
+  withLock(formattersLock):
+    for f in formatters.mitems():
+      testRunEnded(f)
 
+proc ensureInitialized() =
   withLock testFiltersLock:
-    {.gcsafe.}:
-      if not disabledParamFiltering:
-        when declared(paramCount):
-          # Read tests to run from the command line.
-          for i in 1 .. paramCount():
-            testsFilters.incl(paramStr(i))
+    withLock formattersLock:
+      {.gcsafe.}:
+        if not disabledParamFiltering:
+          when declared(paramCount):
+            # Read tests to run from the command line.
+            for i in 1 .. paramCount():
+              let str = paramStr(i)
+              if str.startsWith("--help"):
+                echo "Usage: [--xml=file.xml] [--console] [test-name-glob]"
+                quit 0
+              elif str.startsWith("--xml"):
+                let fn = str[("--xml".len + 1)..^1] # skip separator char as well
+                try:
+                  formatters.add(newJUnitOutputFormatter(
+                    newFileStream(fn, fmWrite)))
+                except CatchableError as exc:
+                  echo "Cannot open ", fn, " for writing: ", exc.msg
+                  quit 1
+              elif str.startsWith("--console"):
+                formatters.add(defaultConsoleFormatter())
+              else:
+                testsFilters.incl(str)
+
+        if formatters.len == 0:
+          # no need for cleanup of console formatter
+          formatters = @[OutputFormatter(defaultConsoleFormatter())]
+        else:
+          addQuitProc(cleanupFormatters)
+
+ensureInitialized() # Run once!
```

Comment:
We can run into trouble with existing test programs that already use command line parameters. This needs to be documented and, just to be on the safe side, provide a define to disable CLI params.

Reply:
well, the current code does something kind of worse, which is to add them as filters and not notify anyone - in addition, it does  so for every single test and suite etc

Reply:
60855d94324e6f27edfbd89cf09751223b67dce5

Reply:
Nice. Should we prefix the define with "unittest" instead of "nimtest"?

Reply:
`unittest2` as prefix if anything, but that would be for a separate PR I think - ie it could make sense to namespace these globals with the project name - I'm not sure this is the best way to do things regardless, ie it would perhaps be better to have a separate import instead (`import unittest2/lite` or the other way around perhaps (`import unittest2/runner`)